flang/lib/Optimizer/Transforms/ArrayValueCopy.cpp

```diff
@@ -116,9 +117,22 @@ class ArrayCopyAnalysis {
 /// Helper class to collect all array operations that produced an array value.
 class ReachCollector {
 public:
-  ReachCollector(llvm::SmallVectorImpl<mlir::Operation *> &reach,
-                 mlir::Region *loopRegion)
-      : reach{reach}, loopRegion{loopRegion} {}
+  /// Return all ops that produce the array value that is stored into the
+  /// `array_merge_store`.
+  static void reachingValues(llvm::SmallVectorImpl<mlir::Operation *> &reach,
+                             ArrayMergeStoreOp mergeStore) {
+    mlir::Value seq = mergeStore.sequence();
+    reach.clear();
+    ReachCollector collector(
+        reach, mlir::cast<ArrayLoadOp>(mergeStore.original().getDefiningOp()),
+        mergeStore);
+    collector.collectArrayMentionFrom(seq);
+  }
+
+private:
+  explicit ReachCollector(llvm::SmallVectorImpl<mlir::Operation *> &reach,
+                          ArrayLoadOp top, ArrayMergeStoreOp bottom)
+      : reach{reach}, exprTop{top}, exprBtm{bottom} {}
```

Comment:
Is this introducing the constraint that first array_load generated in lowering must be for the LHS since this is the one that is used as the top expression boundary here ?

I think at least allocatable assignment needs to lower the RHS before it is dealing with the RHS array_load.